appinventor/components/src/com/google/appinventor/components/runtime/PieChartView.java

```diff
@@ -0,0 +1,253 @@
+      pieChart.notifyDataSetChanged();
+
+      // Invalidate the Chart on the UI thread (via the Handler)
+      // The invalidate method should only be invoked on the UI thread
+      // to prevent exceptions.
+      uiHandler.post(new Runnable() {
+        @Override
+        public void run() {
+          pieChart.invalidate();
+        }
+      });
+    }
+  }
+
+  /**
+   * Resizes all the inner Pie Charts according to the
+   * total count of Pie Charts.
+   */
+  public void resizePieRings() {
+    // Store width and height of last Pie Chart (since getHeight() and
+    // getWidth() will not return the needed result instantly)
+    int lastWidth = 0;
+    int lastHeight = 0;
+
+    float reductionFactor = (0.75f + pieHoleRadius/100f) / pieCharts.size();
+    float radius = (100f - pieHoleRadius);
+    float newRadius = 100f - radius * reductionFactor;
+
+    for (int i = 0; i < pieCharts.size(); ++i) {
+      PieChart pieChart = pieCharts.get(i);
+
+      // Pie Chart non-last: expand radius
+      if (i != pieCharts.size() - 1) {
+        pieChart.setTransparentCircleRadius(newRadius);
+        pieChart.setHoleRadius(newRadius);
+        pieChart.setDrawHoleEnabled(true);
+      } else {
+        float setRadius = pieHoleRadius * (1f + Math.abs(newRadius - pieHoleRadius) / 100f);
+        pieChart.setTransparentCircleRadius(setRadius);
+        pieChart.setHoleRadius(setRadius);
+
+        if (pieHoleRadius == 0) {
+          pieChart.setDrawHoleEnabled(false);
+        }
+      }
+
+      // FP on i != 0 always false
+      if (i != 0) { // Inner Chart
+        // Get the RelativeLayout parameters of the Pie Chart
+        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)(pieChart.getLayoutParams());
+
+        float scalingFactor = (newRadius)/100f;
+
+        // Compute new width & height
+        lastWidth = (int)(lastWidth * scalingFactor);
+        lastHeight = (int)(lastHeight * scalingFactor);
+
+        // Set width & height of the Pie Chart, and update the Layout parameters
+        // of the Chart
+        params.width = lastWidth;
+        params.height = lastHeight;
+        pieChart.setLayoutParams(params);
+      } else { // Root Chart
+        // Set last height & width
+        lastHeight = pieChart.getHeight();
+        lastWidth = pieChart.getWidth();
+      }
+
+      // Since the Legend is drawn inside, an offset is needed
+      // so that the rings of the Pie Chart do not overlap with the
+      // Legend. The value was chosen as an optimal value (10 is too large,
+      // 5 is too small, 7 is a somewhat in the middle option)
+      // TODO: This could be improved in the future to (perhaps) dynamically
+      // TODO: adjust the bottom offset.
+      pieChart.setExtraBottomOffset(7);
+    }
+  }
+
+  /**
+   * Adds a new Legend Entry to the Legend of the Pie Chart view.
+   * @param entry  Legend Entry to add
+   */
+  public void addLegendEntry(LegendEntry entry) {
```

Comment:
Additional tests for Legend Entry could be done:
- [ ] Adding from inner Pie Charts
- [ ] Remove non-existent entries